### PR TITLE
Storage API and async bigquery request

### PR DESCRIPTION
**Pull request by alasarr**
Related to https://github.com/CartoDB/cartoframes/issues/1257

- [x] Download using storage API with fallback to default API in case of exception
- [x] Async operations to BigQuery. In the previous version, if we had multiple datasets each request was not done in parallel
- [x] Time profile as a decorator. To check with the team if this approach is feasible

- [x] Unit test

- [x] Use token instead of the service account (hard-coded)

- [x] Adapt to download

- [ ] Ask Google about a stable version of Storage API

### CODE REVIEW

**Comment by oleurud** [comment]
I am getting this error using BQ storage API: `400 request failed: the table has a storage format that is not supported` I have not investigated it. I am changing to another task

**Comment by alasarr** [comment]
>but it seems to be related with tables smaller than 10MB (in this case, the fallback will be required). I am changing to another task
Yeap, that's the reason why I included the fallback

**Comment by oleurud** [comment]
Ready to review and acceptance

**Comment by alasarr** [comment]
@oleurud, I've fixed the following things:

-  `__get_geo_table` now check if the geography is_public_data. Before, it just checked the dataset and it doesn't work like data. AGS has a geography that is public_data.

- If a user doesn't have a subscription the method `_is_subscribed` failed.

- Infinite loop when a query had some errors.

It's not the goal of this task, but tests need a refactor, it doesn't make too much sense the amount of time rewrite a test requires (too much copy&paste)

File: cartoframes/data/observatory/enrichment/enrichment_service.py [hunk 1]
```diff
@@ -175,9 +183,10 @@ def __get_dataset(self, variable, table_name):
 
     def __get_geo_table(self, variable):
         geography_id = Dataset.get(variable.dataset).geography
+        geography = Geography.get(geography_id)
         _, dataset_geo_table, geo_table = geography_id.split('.')
 
-        if variable.project_name != self.public_project:
+        if not geography.is_public_data:
```
Comment (oleurud) [inline cartoframes/data/observatory/enrichment/enrichment_service.py]: I like it, easier to understand, but was it failing?

Comment (oleurud) [reply cartoframes/data/observatory/enrichment/enrichment_service.py]: Ahhh, right, it was a bug. Good catch!

**Comment by oleurud** [comment]
We need to merge https://github.com/CartoDB/cartoframes/pull/1311 first, and then prepare this one with the changes

**Comment by oleurud** [comment]
Ready again! CR and acceptance

File: cartoframes/data/clients/bigquery_client.py [hunk 2]
```diff
@@ -36,7 +40,9 @@ def __init__(self, credentials):
         self._credentials = credentials or get_default_credentials()
         self.bq_client = None
         self.gcs_client = None
+        self.bq_storage_client = None
 
+        self._gcp_execution_project = None
```
Comment (Jesus89) [inline cartoframes/data/clients/bigquery_client.py]: Why `_` here?

File: cartoframes/data/clients/bigquery_client.py [hunk 3]
```diff
@@ -89,58 +163,29 @@ def upload_dataframe(self, dataframe, schema, tablename):
             uri, table_ref, job_config=job_config
         )
 
-        job.result()  # Waits for table load to complete.
```
Comment (Jesus89) [inline cartoframes/data/clients/bigquery_client.py]: We can keep this comment IMO

File: cartoframes/data/clients/bigquery_client.py [hunk 4]
```diff
@@ -89,58 +163,29 @@ def upload_dataframe(self, dataframe, schema, tablename):
...
-        return [field.name for field in table_info.schema]
 
-    def download_to_file(self, project, dataset, table, file_path=None, limit=None, offset=None,
-                         fail_if_exists=False, progress_bar=True):
-        if not file_path:
-            file_name = '{}.{}.{}.csv'.format(project, dataset, table)
-            file_path = os.path.join(_USER_CONFIG_DIR, file_name)
+def _rows_to_file(rows, file_path, column_names=None, progress_bar=True):
+    if progress_bar:
+        pb = tqdm.tqdm_notebook(total=rows.total_rows)
```
Comment (Jesus89) [inline cartoframes/data/clients/bigquery_client.py]: What happens if you run this from a console?

File: setup.py [hunk 5]
```diff
@@ -33,6 +33,8 @@ def get_version():
     'pyarrow>=0.14.1,<1.0',
     'google-cloud-storage>=1.23.0,<2.0',
     'google-cloud-bigquery>=1.22.0,<2.0',
+    'google-cloud-bigquery-storage>=0.7.0,<1.0',
+    'fastavro>=0.22.7,<1.0',
```
Comment (Jesus89) [inline setup.py]: Why is this `fastavro` required?

File: tests/unit/data/observatory/enrichment/test_polygon_enrichment.py [hunk 6]
```diff
@@ -26,13 +32,10 @@ def __init__(self, public_data_project, user_data_project, access_token='access_
...
 
-class CatalogEntityWithGeographyMock:
-    def __init__(self, geography):
-        self.geography = geography
-
-
 class TestPolygonEnrichment(object):
     def setup_method(self):
+        # self.original_init_clients = BigQueryClient._init_clients
+        # BigQueryClient._init_clients = Mock(return_value=(True, True, True))
```
Comment (Jesus89) [inline tests/unit/data/observatory/enrichment/test_polygon_enrichment.py]: We should remove commented code

Comment (oleurud) [reply tests/unit/data/observatory/enrichment/test_polygon_enrichment.py]: Ups

File: setup.py (hunk 5 above)
Comment (oleurud) [reply setup.py]: It is required by BigQuery Storage API

File: cartoframes/data/clients/bigquery_client.py (hunk 4 above)
Comment (oleurud) [reply cartoframes/data/clients/bigquery_client.py]: No idea. I will check if fails

File: cartoframes/data/clients/bigquery_client.py (hunk 2 above)
Comment (oleurud) [reply cartoframes/data/clients/bigquery_client.py]: It is only used internally

File: cartoframes/data/clients/bigquery_client.py (hunk 3 above)
Comment (oleurud) [reply cartoframes/data/clients/bigquery_client.py]: Ok

**Pull request merged**